### PR TITLE
[Feature] Add trajectory recording to BAGEL denoising loop

### PR DESCRIPTION
**Pull request by timzsu**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Enable opt-in capture of intermediate latents and timesteps during BAGEL's flow-matching denoising process, providing rollout data needed by RL algorithms like FlowGRPO. Part of RFC #1904.

**Denoising loop** (`bagel_transformer.py`): Instrument all 4 loop variants (batched CFG, SP+CFG, SP-only, CFG-parallel) with `return_trajectory_latents` flag. When enabled, clone `x_t` and record the timestep after each Euler step. Log probs will also be recorded if there is a scheduler. Zero overhead when disabled (default).

**Pipeline** (`pipeline_bagel.py`): Read `return_trajectory_latents` / `return_trajectory_decoded` from `DiffusionSamplingParams`. Stack trajectory latents, timesteps, and log-probs (if any); optionally, VAE-decode each intermediate latent to a PIL image.

**Output plumbing** (`outputs.py`, `diffusion_engine.py`, `stage_diffusion_proc.py`, `omni_base.py`): Add `trajectory_latents`, `trajectory_timesteps`, `trajectory_log_probs`, `trajectory_decoded` fields to `OmniRequestOutput` and propagate them through the diffusion engine, the ZMQ-based inter-process merge in `StageDiffusionProc`, and the final `OmniBase` output construction.

**Tests** (`test_trajectory_recording.py`): 12 CPU-only unit tests exercising the no-CFG/CFG code path via a lightweight mock. Plus 4 CPU-only unit tests that check the trajectory log-probs via a mock scheduler (a real one will be introduced in #2501). 

## Test Plan

Unit tests that mock the generation phase and only test the trajectory return. 
```bash
python -m pytest tests/diffusion/models/bagel/test_trajectory_recording.py -v
```

The code is also tested locally with an end-to-end inference script. But I think it is too heavy to include an e2e test script in this lightweight PR. Feel free to request that I add it. 

## Test Result

All 16 unit tests passed. 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by timzsu** [comment]
@princepride Could you please take a look at this PR?

**Comment by princepride** [comment]
@SamitHuang I've noticed a difference in the trajectory implementation between Bagel and Qwen-Image.

In my opinion, for small experiments, single-process setups, and small tensors: BAGEL can just use custom_output and modify the pipeline, similar to the Qwen tests.
However, the more robust approach is to use structured fields (trajectory_latents / timesteps / decoded) and ensure they are recognized end-to-end across IPC and OmniRequestOutput, rather than just stuffing them into custom_output. Do we need to update our Qwen-Image implementation?

File: tests/diffusion/models/bagel/test_trajectory_recording.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    def test_trajectory_final_latent_matches_output(self, bagel_and_args):
+        bagel, args = bagel_and_args
+
+        unpacked, trajectory_latents, _ = bagel.generate_image(**args, return_trajectory_latents=True)
+
+        # Reconstruct the full final latent from unpacked pieces
+        final_latent = torch.cat(unpacked, dim=0)
+        assert torch.allclose(trajectory_latents[-1], final_latent, atol=1e-6), (
+            "Last trajectory latent should match the final output"
+        )
```
Comment (hsliuustc0106) [inline tests/diffusion/models/bagel/test_trajectory_recording.py:164]: These 6 tests provide good coverage for the no-CFG path, but the PR modifies 4 denoising loop variants (batched CFG, SP+CFG, SP-only, and CFG-parallel). Consider adding at least one test with `cfg_text_scale > 1.0` to verify trajectory recording works correctly in the CFG code paths.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 2]
```diff
@@ -1756,9 +1762,12 @@ def generate_image(
                     )
 
                 x_t = x_t - v_t.to(x_t.device) * dts[i]
+                if return_trajectory_latents:
+                    trajectory_latents.append(x_t.clone())
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:1780]: This collects individual scalar tensors (0-dim) into a list. For consistency with `trajectory_latents` (which gets stacked), consider either:
1. Converting to a 1D tensor: `trajectory_timesteps = torch.tensor(trajectory_timesteps)` after the loop, or
2. Documenting that this intentionally returns a list of scalar tensors

The current approach works but may surprise users expecting a stacked tensor like `trajectory_latents`.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 3]
```diff
@@ -650,11 +650,27 @@ def vae_transforms(img):
                 cfg_img_packed_query_indexes=generation_input_cfg_img["cfg_packed_query_indexes"],
                 cfg_img_key_values_lens=generation_input_cfg_img["cfg_key_values_lens"],
                 cfg_img_packed_key_value_indexes=generation_input_cfg_img["cfg_packed_key_value_indexes"],
+                return_trajectory_latents=req.sampling_params.return_trajectory_latents,
             )
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:659]: Good: `trajectory_latents` is properly stacked into a tensor. For consistency, consider also converting `trajectory_timesteps` to a tensor here (e.g., `trajectory_timesteps = torch.tensor(trajectory_timesteps) if trajectory_timesteps else None`).

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 4]
```diff
@@ -650,11 +650,27 @@ def vae_transforms(img):
...
+                trajectory_decoded = [
+                    self._decode_image_from_latent(self.bagel, self.vae, lat, image_shape) for lat in trajectory_latents
+                ]
+
         return DiffusionOutput(
-            output=img, stage_durations=self.stage_durations if hasattr(self, "stage_durations") else None
+            output=img,
+            trajectory_latents=trajectory_latents_stacked,
+            trajectory_timesteps=trajectory_timesteps,
+            trajectory_decoded=trajectory_decoded,
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:684]: I don't see the necessity of returning `trajectory_decoded`, it's more common to return `trajectory_log_probs` for FlowGRPO algo with SDE Scheduler. #2501 also adopts standard FlowGRPO algo

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 5]
```diff
@@ -650,11 +650,27 @@ def vae_transforms(img):
...
                 cfg_img_key_values_lens=generation_input_cfg_img["cfg_key_values_lens"],
                 cfg_img_packed_key_value_indexes=generation_input_cfg_img["cfg_packed_key_value_indexes"],
+                return_trajectory_latents=req.sampling_params.return_trajectory_latents,
             )
 
         img = self._decode_image_from_latent(self.bagel, self.vae, latents[0], image_shape)
+
+        # Build trajectory output when requested
+        trajectory_latents_stacked = None
+        trajectory_decoded: list[torch.Tensor] | None = None
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py]: `self._decode_image_from_latent` returns a PIL `Image.Image`, but this is annotated as `list[torch.Tensor]`. Please update the annotation to `list[Image.Image] | None = None` (and ensure PIL is imported if needed, or simply omit the type hint).

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 6]
```diff
@@ -650,11 +650,27 @@ def vae_transforms(img):
...
+            if req.sampling_params.return_trajectory_decoded:
+                trajectory_decoded = [
+                    self._decode_image_from_latent(self.bagel, self.vae, lat, image_shape) for lat in trajectory_latents
+                ]
+
         return DiffusionOutput(
-            output=img, stage_durations=self.stage_durations if hasattr(self, "stage_durations") else None
+            output=img,
+            trajectory_latents=trajectory_latents_stacked,
+            trajectory_timesteps=trajectory_timesteps,
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py]: Please stack the timesteps here as well, similar to how latents are stacked. Downstream RL rollout buffers will expect a 1D tensor of shape `(num_steps,)` rather than a Python list. e.g., `trajectory_timesteps_stacked = torch.stack(trajectory_timesteps) if trajectory_timesteps else None`.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
small fixes

**Comment by SamitHuang** [comment]
> @SamitHuang I've noticed a difference in the trajectory implementation between Bagel and Qwen-Image.
> 
> In my opinion, for small experiments, single-process setups, and small tensors: BAGEL can just use custom_output and modify the pipeline, similar to the Qwen tests. However, the more robust approach is to use structured fields (trajectory_latents / timesteps / decoded) and ensure they are recognized end-to-end across IPC and OmniRequestOutput, rather than just stuffing them into custom_output. Do we need to update our Qwen-Image implementation?

good suggestion. we used `custom_output` for quick prototyping and to allow flexible return data format required by different RL algorithms, but sending arbitrary dict containing large tensors across the IPC layer using ZMQ can easily break or cause silent serialization bugs. We can refactor qwen-image with `trajectory_*` fields for structured outputs.

But for the API design, we should determine whether we allow adding each item to `DiffusionOutput` required by different algo(SRPO, FlowGRPO, FlowGRPO-fast), currently we have these items: log_prob, latent, timesteps.

File: tests/diffusion/models/bagel/test_trajectory_recording.py (hunk 1 above)
Comment (timzsu) [reply tests/diffusion/models/bagel/test_trajectory_recording.py:164]: Added another 6 tests with cfg_text_scale = 4.0

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py (hunk 2 above)
Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1780]: It is stacked later, as discussed by the next comment.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py (hunk 3 above)
Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py:659]: Done. timesteps are also stacked.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py (hunk 5 above)
Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py]: Fixed.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py (hunk 6 above)
Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py]: This is the same issue as pointed out by @hsliuustc0106 . Fixed.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py (hunk 4 above)
Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py:684]: I think this switch is useful for debugging. When I tested the end-to-end workflow, the decoded images form a clear chain from noise to real image.

Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py:684]: I have added log probs in trajectory. Note that since there is no scheduler, the log prob is ineffective in this PR. #2501 introduces the SDE scheduler, and then this trajectory will be used.

**Comment by SamitHuang** [comment]
@codex review

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 7]
```diff
@@ -1778,10 +1801,19 @@ def generate_image(
                     past_key_values=past_key_values,
                     packed_key_value_indexes=packed_key_value_indexes,
                 )
-                x_t = x_t - v_t.to(x_t.device) * dts[i]
+                if scheduler is not None:
+                    out = scheduler.step(v_t.to(x_t.device), timesteps[i], x_t, dts[i], **_sched_kw)
+                    x_t = out.prev_sample
+                    if trajectory_log_probs is not None and out.log_prob is not None:
+                        trajectory_log_probs.append(out.log_prob)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]:
```suggestion
                    out_log_prob = getattr(out, "log_prob", None)
                    if trajectory_log_probs is not None and out_log_prob is not None:
                        trajectory_log_probs.append(out_log_prob)
```

Comment (SamitHuang) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py]: some scheduler output may not have log_prob attribute, check before accessing to avoid crashing

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 8]
```diff
@@ -650,11 +653,36 @@ def vae_transforms(img):
...
+        trajectory_log_probs_stacked: torch.Tensor | None = None
+        if trajectory_log_probs:
+            trajectory_log_probs_stacked = torch.stack(trajectory_log_probs)
+
         return DiffusionOutput(
-            output=img, stage_durations=self.stage_durations if hasattr(self, "stage_durations") else None
+            output=img,
+            trajectory_latents=trajectory_latents_stacked,
+            trajectory_timesteps=trajectory_timesteps_stacked,
+            trajectory_log_probs=trajectory_log_probs_stacked,
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:683]: @SamitHuang @timzsu I think `trajectory_log_probs` may also need add to `vllm_omni/diffusion/ipc.py::_pack_diffusion_fields`, because it may also too big to serialize inline for zmq path.

Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py:683]: @princepride Updated. But I've noticed a pre-existing typing issue in `ipc.py`. The packed diffusion output could be of type `dict`, yet the typing annotations only show `torch.Tensor | None`. There are two ways to fix: 

1. Declare a `PackedDiffusionOutput` class to carry the packed output; or
2. Make `DiffusionOutput` accept `torch.Tensor | dict | None`.

Which one should we choose? Should it be in this PR or in a separate one?

Comment (timzsu) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py:683]: Fixed using the second way, which I think has a smaller impact on the codebase.
